### ISSUE TITLE
Producer stops working after loss of communication to Kafka

### ISSUE THREAD

**Issue by smueller18** [Parsely/pykafka#629]
**PyKafka version**: 2.6.0.dev1
**Kafka version**: 0.9.0.0

I encountered a problem by using a pykafka producer if Kafka is shutdown during script execution and restarted again little while later. Here is the description in which order steps have to be executed.
1. Kafka is running appropriately
2. Python producer script is started
3. Terminate Kafka process
4. After 3 connection retries, Python script logs "Worker exited for broker ..." but script is still running
5. Start Kafka again
6. Python script does not create a new produce worker

Here is my little python script:
```python
import time
import logging.config
from pykafka import KafkaClient

LOGGING_LEVEL = "DEBUG"
KAFKA_HOSTS = "localhost:9092"
TOPIC = "mytopic"
PUSH_INTERVAL = 5

logging.basicConfig(level=LOGGING_LEVEL)
client = KafkaClient(hosts=KAFKA_HOSTS)
topic = client.topics[str.encode(TOPIC)]

producer = topic.get_producer(use_rdkafka=False, sync=True)
while True:
    producer.produce(str.encode(str(time.time())))
    time.sleep(PUSH_INTERVAL)
```
And the logging output:
```
[...]
DEBUG:pykafka.producer:Sending 1 messages to broker 0
WARNING:pykafka.producer:Broker b'172.19.0.3':9092 disconnected. Retrying.
DEBUG:pykafka.cluster:Updating cluster, attempt 1/3
DEBUG:pykafka.connection:Connecting to localhost:9092
ERROR:pykafka.connection:Failed to connect to localhost:9092
WARNING:pykafka.broker:Encountered SocketDisconnectedError while requesting metadata from broker localhost:9092. Continuing.
DEBUG:pykafka.broker:Retrying
WARNING:pykafka.broker:Encountered SocketDisconnectedError while requesting metadata from broker localhost:9092. Continuing.
DEBUG:pykafka.broker:Retrying
WARNING:pykafka.broker:Encountered SocketDisconnectedError while requesting metadata from broker localhost:9092. Continuing.
DEBUG:pykafka.connection:Connecting to localhost:9092
ERROR:pykafka.connection:Failed to connect to localhost:9092
WARNING:pykafka.broker:Encountered SocketDisconnectedError while requesting metadata from broker localhost:9092. Continuing.
DEBUG:pykafka.broker:Retrying
WARNING:pykafka.broker:Encountered SocketDisconnectedError while requesting metadata from broker localhost:9092. Continuing.
DEBUG:pykafka.broker:Retrying
WARNING:pykafka.broker:Encountered SocketDisconnectedError while requesting metadata from broker localhost:9092. Continuing.
DEBUG:pykafka.connection:Connecting to localhost:9092
ERROR:pykafka.connection:Failed to connect to localhost:9092
WARNING:pykafka.broker:Encountered SocketDisconnectedError while requesting metadata from broker localhost:9092. Continuing.
DEBUG:pykafka.broker:Retrying
WARNING:pykafka.broker:Encountered SocketDisconnectedError while requesting metadata from broker localhost:9092. Continuing.
DEBUG:pykafka.broker:Retrying
WARNING:pykafka.broker:Encountered SocketDisconnectedError while requesting metadata from broker localhost:9092. Continuing.
WARNING:pykafka.producer:No brokers available
DEBUG:pykafka.producer:Successfully sent 0/1 messages to broker 0
INFO:pykafka.producer:Worker exited for broker b'172.19.0.3':9092
```
After the last log message, the script is still running but does not react if Kafka is available again.
In my case, this is fatal, because Kafka is not running stable or some ethernet connection errors can appear. For this reason I wrote a workaround. But in my eyes, this is very poor. I use a logging Filter which searches for "Encountered SocketDisconnectedError"-messages and the script sends a SIGINT signal to itself on matches:
```python
import time
import os
import signal
import logging.config
from pykafka import KafkaClient

LOGGING_LEVEL = "DEBUG"
KAFKA_HOSTS = "localhost:9092"
TOPIC = "mytopic"
PUSH_INTERVAL = 5

logging.basicConfig(level=LOGGING_LEVEL)
client = KafkaClient(hosts=KAFKA_HOSTS)
topic = client.topics[str.encode(TOPIC)]

logger = logging.getLogger('producer')
class ListenFilter(logging.Filter):
    def filter(self, record):
        if record.getMessage().startswith("Encountered SocketDisconnectedError"):
            logger.error("Lost connection to Kafka")
            os.kill(os.getpid(), signal.SIGINT)
        return True
    
logging.getLogger('pykafka.broker').addFilter(ListenFilter())

producer = topic.get_producer(use_rdkafka=False, sync=True)
while True:
    producer.produce(str.encode(str(time.time())))
    time.sleep(PUSH_INTERVAL)
```

Is it possible to guarantee a recreation of Kafka producer or at least throw a `NoBrokersAvailableError` exception like it is done if Kafka is not running on script startup?

**Comment by emmett9001** [comment]
Thanks for the report, @smueller18. I gather that you're doing your tests with a single-node Kafka cluster such that there are no brokers available when you shut down the broker at `localhost:9092`. If that's the case, I agree that the producer should do something other than hanging in this situation.

**Comment by smueller18** [comment]
Yes @emmett9001, your assumption of single-node Kafka cluster is right. But additionally, on network breakdowns between producers and the Kafka cluster (also multi-node) the behavier of pykafka would be the same.

**Comment by emmett9001** [comment]
What appears to be happening here is that all of the producer's worker threads are being stopped by [this logic](https://github.com/Parsely/pykafka/blob/aa30424d68ad03ae4e702c0fb95e5e9650058ae4/pykafka/producer.py#L231) and have [no facility](https://github.com/Parsely/pykafka/blob/aa30424d68ad03ae4e702c0fb95e5e9650058ae4/pykafka/producer.py#L533) for periodically reattempting the connection. In situations where some `n < N` of the nodes in an `N`-node cluster fail such that its leadership assignments are changed, this is handled by [the worker's response to `NotLeaderForPartition`](https://github.com/Parsely/pykafka/blob/aa30424d68ad03ae4e702c0fb95e5e9650058ae4/pykafka/producer.py#L433), which results in a call to `_update()`. Note that this failure mode is distinct from the nodes becoming unreachable such that their leadership assignments are not changed - I think that case probably also exhibits this issue. When all `N` of the nodes become unreachable, the worker threads for these nodes have all been stopped and are not reopened. This doesn't hang the main thread, but it does mean that any future calls to `produce()` will result in no data being sent to brokers.

If this kind of failure could only happen as a result of all brokers going dark, I'd say it would be enough to simply remove [this exception handler](https://github.com/Parsely/pykafka/blob/aa30424d68ad03ae4e702c0fb95e5e9650058ae4/pykafka/producer.py#L451) and allow the exception to bubble up to the main thread. However, the issue of a broker's worker thread dying and never being recreated can theoretically occur even for a subset of the nodes in a cluster. For this to happen, the nodes have to become unreachable but retain their leadership assignments such that `NotLeaderForPartition` is never returned to the producer - conceivable in a degraded network environment.

The question, then, is how exactly to fix this. The low-hanging fruit is to allow `NoBrokersAvailable` to bubble to the main thread. The more complete fix would involve periodically attempting to restart `OwnedBroker`s that had been stopped previously - probably by calling `_setup_owned_brokers`.

**Comment by emmett9001** [comment]
Removing the `NoBrokersAvailable` exception handler seems like the way to go. See my comment on #449. I was a bit off on the idea that nodes becoming unreachable could also put the cluster into this state, since any such situation will result in a call to `_update`.

**Comment by emmett9001** [comment]
@smueller18 Please give #633 a try and let me know if you follow my logic / whether it works for you.

**Comment by smueller18** [comment]
@emmett9001 thanks for the fix. #633 exactly fixes this issue (#629) by throwing NoBrokersAvilableError to main thread. Validated with single-node Kafka cluster only.
